### PR TITLE
ring: modify `DoUntilQuorum` to optionally only send the minimum number of requests required to reach quorum

### PR DESCRIPTION
**Pull request by charleskorn**
**What this PR does**:

This PR modifies `DoUntilQuorum` to optionally only send the minimum number of requests required to reach quorum. 

For example, assume zone-awareness is enabled, the replication set contains three zones and one zone failure is tolerated:

* With request minimization disabled, `DoUntilQuorum` behaves as it does currently: it calls `f` for all instances in the replication set immediately, but only returns the results from the calls required to reach quorum and cancels any outstanding calls once quorum is reached.

* With request minimization enabled, `DoUntilQuorum` would initially only call `f` for instances in two of the three zones. If those calls all succeed, then `DoUntilQuorum` returns those results and ignores the remaining zone. If any call fails, then `DoUntilQuorum` calls `f` for the instances in the remaining zone. The set of results returned only includes results from instances in successful zones, as before.

Enabling request minimization is expected to reduce the number of calls made in the happy case (all requests succeed) at the cost of additional latency in the event that one or more of the initial calls fails.

This is a breaking change to the signature for `DoUntilQuorum`. As far as I can see, `DoUntilQuorum` is only used in two places, once in Mimir ([source](https://github.com/grafana/mimir/blob/bea4bd616c42c5ea42b3cbc7e350e3ebf2a291f0/pkg/distributor/query.go#L306)) and once in Phlare ([source](https://github.com/grafana/phlare/blob/d9eb816015a8d255c8faa4c58a4f4f169a3cabc1/pkg/querier/replication.go#L29)), so the small impact seems worthwhile over introducing a second method to maintain backwards compatibility.

**Which issue(s) this PR fixes**:

(none)

**Checklist**
- [x] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: ring/replication_set.go [hunk 1]
```diff
@@ -140,11 +154,27 @@ func DoUntilQuorum[T any](ctx context.Context, r ReplicationSet, f func(context.
...
+	} else {
+		resultTracker.releaseAllRequests()
+	}
+
 	for i := range r.Instances {
 		instance := &r.Instances[i]
 		instanceCtx := contextTracker.contextFor(instance)
 
 		go func(desc *InstanceDesc) {
+			if !resultTracker.awaitRelease(desc) {
```
Comment (pracucci) [inline ring/replication_set.go:168]: Below, when the context is canceled we call `cleanupResultsAlreadyReceived()` and return. I've the feeling we may have a condition where this goroutine is never terminated because `awaitRelease()` waits indefinitely in that case.

[non blocking] Generally speaking, I've found quite hard to follow the "failure" case to make sure `awaitRelease()` always return in every case. I'm wondering if there's anything we can do to make the code easier to follow.

**Review by pracucci — COMMENTED** [review]
Good job! I find the overall logic non trivial to follow, but I also don't have a good suggestion to improve it, so I wouldn't block on it. The reason why I'm not approving yet is because I have a doubt about the context cancellation case (see related comment) I would like you to double check. Thanks!

File: ring/replication_set_tracker.go [hunk 2]
```diff
@@ -20,6 +25,25 @@ type replicationSetResultTracker interface {
 	// This method should only be called after succeeded returns true for the first time and before
 	// calling done any further times.
 	shouldIncludeResultFrom(instance *InstanceDesc) bool
+
+	// Releases an initial set of requests sufficient to meet the quorum requirements of this tracker.
+	// Further requests will be released if necessary when done is called with a non-nil error.
+	// Calling this method multiple times may lead to unpredictable behaviour.
+	// Calling both this method and releaseAllRequests may lead to unpredictable behaviour.
+	// This method must only be called before calling done.
+	releaseMinimumRequests()
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply ring/replication_set_tracker.go]: If "start" is clearer to you, then I'm happy to change it.

I've done this in b9a4db9.

File: ring/replication_set.go (hunk 1 above)
Comment (charleskorn) [reply ring/replication_set.go]: > Below, when the context is canceled we call `cleanupResultsAlreadyReceived()` and return. I've the feeling we may have a condition where this goroutine is never terminated because `awaitRelease()` waits indefinitely in that case.

Good catch, yep, that could block forever. I'll add a test for this case and then fix it.

> [non blocking] Generally speaking, I've found quite hard to follow the "failure" case to make sure `awaitRelease()` always return in every case. I'm wondering if there's anything we can do to make the code easier to follow.

We might be able to solve both of these problems in one go: I'll modify `awaitRelease()` to take the request context as well, so if that is cancelled, it'll return immediately. This should make it easier to confirm that `awaitRelease()` will always return.

Comment (charleskorn) [reply ring/replication_set.go]: I've implemented this (f272c76), however it introduces additional complexity for the tests (93656f2):

Because `awaitStart()` (formerly `awaitRelease()`) now returns if the instance's context has been cancelled, there's a chance that `f` is not always called and therefore won't need to be cleaned up. The tests previously assumed this was more deterministic, and so the assertions around cleanup behaviour are now more complex.
